HPCC-21779 Std.File.Copy does not have nosplit option

AttilaVamos opened the pull request:
Add nosplit option to:
- Copy
- SprayFixed
- SprayVariable
- SprayXml
- RemotePull
functions.

Handle new parameter in fileservices.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

Tested manually.

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-12836, label: HPCC-21779-fix-7.4.x starts now.
The reason of this test is: Did not build yet.
Commit ID: 7334f44cdb8a3c62930f15a5e79cea6f71a09a2e
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 7334f44cdb8a3c62930f15a5e79cea6f71a09a2e
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |17 sec  (00:00:17) |
|test (hthor) |847 |847 |0 |198 sec  (00:03:18) |
|test (thor) |770 |770 |0 |709 sec  (00:11:49) |
|test (roxie) |923 |923 |0 |230 sec  (00:03:50) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  249 sec (00:04:09) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1468 sec (00:24:28) |   20 sec (00:00:20) | 1793 sec (00:29:53) |

HPCCSmoketest commented:
Process of PR-12836, label: HPCC-21779-fix-7.4.x starts now.
The reason of this test is: New commit.
Commit ID: abd83078722878ba906857869a115438dd13458e
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: abd83078722878ba906857869a115438dd13458e
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |26 sec  (00:00:26) |
|test (hthor) |847 |847 |0 |184 sec  (00:03:04) |
|test (thor) |771 |771 |0 |690 sec  (00:11:30) |
|test (roxie) |924 |924 |0 |227 sec  (00:03:47) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   33 sec (00:00:33) |  252 sec (00:04:12) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1419 sec (00:23:39) |   20 sec (00:00:20) | 1745 sec (00:29:05) |

HPCCSmoketest commented:
This PR (12836) is in the 1/2 position of the queue. Estimated start time is after ~0.50 hour(s)

HPCCSmoketest commented:
Process of PR-12836, label: HPCC-21779-fix-7.4.x starts now.
The reason of this test is: New commit.
Commit ID: 765616f8297714425eda0b7a66d075a9f1f7650d
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 765616f8297714425eda0b7a66d075a9f1f7650d
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |847 |847 |0 |174 sec  (00:02:54) |
|test (thor) |771 |771 |0 |684 sec  (00:11:24) |
|test (roxie) |923 |923 |0 |259 sec  (00:04:19) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   38 sec (00:00:38) |  255 sec (00:04:15) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1432 sec (00:23:52) |   19 sec (00:00:19) | 1765 sec (00:29:25) |

HPCCSmoketest commented:
Process of PR-12836, label: HPCC-21779-fix-7.4.x starts now.
The reason of this test is: New commit.
Commit ID: 765616f8297714425eda0b7a66d075a9f1f7650d
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 765616f8297714425eda0b7a66d075a9f1f7650d
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |26 sec  (00:00:26) |
|test (hthor) |847 |847 |0 |208 sec  (00:03:28) |
|test (thor) |771 |771 |0 |689 sec  (00:11:29) |
|test (roxie) |923 |923 |0 |417 sec  (00:06:57) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   27 sec (00:00:27) |  249 sec (00:04:09) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1620 sec (00:27:00) |   20 sec (00:00:20) | 1938 sec (00:32:18) |

AttilaVamos commented:
@jakesmith please review.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2054,25 +2106,8 @@ FILESERVICES_API void FILESERVICES_CALL fsRemotePull(ICodeContext *ctx,
-                                                     bool wrap,
-                                                     bool compress)
+                                                     bool compress,
+                                                     bool noSplit)
```
should this have expiryDays too?

AttilaVamos replied:
Yes

jakesmith commented on testing/regress/ecl/split_test.ecl:
```diff
@@ -0,0 +1,574 @@
+allPeople := DATASET([ {'foo', 10, 1},
+                       {'bar', 12, 0},
+                       {'baz', 32, 1}]
+            ,Layout_Person);
```
Would be better test if initial set was larger and distributed.
Could do something like:
```
allPeople := DATASET(10000, TRANSFORM(layout_Person, SELF.name := (string)COUNTER; SELF.age := COUNTER%99; SELF.good := IF(COUNTER%2=1, TRUE, FALSE)), DISTRIBUTED);
```

AttilaVamos replied:
Yes, but I should test it with SprayFixed().

AttilaVamos replied:
Changed and updated for SprayFix().

jakesmith commented:
@AttilaVamos - looks good, a couple of minor comments.

HPCCSmoketest commented:
Process of PR-12836, label: HPCC-21779-fix-7.4.x starts now.
The reason of this test is: New commit.
Commit ID: aed70355ce6c872d3858923f2601a07fd277f331
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: aed70355ce6c872d3858923f2601a07fd277f331
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |848 |848 |0 |186 sec  (00:03:06) |
|test (thor) |772 |772 |0 |695 sec  (00:11:35) |
|test (roxie) |924 |924 |0 |234 sec  (00:03:54) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  255 sec (00:04:15) |    0 sec (00:00:00) |    4 sec (00:00:04) |   19 sec (00:00:19) | 1433 sec (00:23:53) |   21 sec (00:00:21) | 1766 sec (00:29:26) |

AttilaVamos commented:
@jakesmith  Updated, please re-review.

jakesmith commented:
@AttilaVamos - looks good, please squash.
Should this be 7.6?

AttilaVamos committed:
HPCC-21779 Std.File.Copy does not have nosplit option

Add nosplit option to:
- Copy
- SprayFixed
- SprayVariable
- SprayXml
- RemotePull
functions.

Create a new Regression Suite case split_test.ecl to test nosplit paramter.

Add expire parameter to Copy function

Extend spray_exipre_test.ecl to test std.file.copy().

Handle new paramter in fileservices.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Process of PR-12836, label: HPCC-21779-fix-7.4.x starts now.
The reason of this test is: New commit.
Commit ID: 4b3e2672b71369750712794cff61fef385a01b7c
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 4b3e2672b71369750712794cff61fef385a01b7c
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |851 |851 |0 |204 sec  (00:03:24) |
|test (thor) |775 |775 |0 |716 sec  (00:11:56) |
|test (roxie) |927 |927 |0 |245 sec  (00:04:05) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   40 sec (00:00:40) |  250 sec (00:04:10) |    0 sec (00:00:00) |    4 sec (00:00:04) |   19 sec (00:00:19) | 1479 sec (00:24:39) |   20 sec (00:00:20) | 1812 sec (00:30:12) |

AttilaVamos commented:
@richardkchapman Should this be 7.6.x?

AttilaVamos commented:
@jakesmith squashed.

richardkchapman commented:
@AttilaVamos Yes, I would say 7.6 feels more appropriate

AttilaVamos commented:
@jakesmith Close to re-target. The new PR is: https://github.com/hpcc-systems/HPCC-Platform/pull/12928
